crates/typst/src/eval/plugin.rs

```diff
@@ -0,0 +1,204 @@
+use crate::diag::SourceResult;
+use crate::util::Bytes;
+use ecow::EcoString;
+use std::sync::Mutex;
+use typst::diag::At;
+use wasmi::{AsContext, Caller, Engine, Func as Function, Linker, Module, Value};
+
+#[derive(Debug)]
+pub struct Plugin(Mutex<PluginInstance>, Bytes);
+
+impl crate::eval::Type for Plugin {
+    const TYPE_NAME: &'static str = "Plugin";
+}
+
+impl PartialEq for Plugin {
+    fn eq(&self, other: &Self) -> bool {
+        self.1 == other.1
+    }
+}
+
+impl std::hash::Hash for Plugin {
+    fn hash<H: std::hash::Hasher>(&self, state: &mut H) {
+        self.1.hash(state);
+    }
+}
+
+impl Plugin {
+    pub fn new(instance: PluginInstance, bytes: Bytes) -> Self {
+        Self(Mutex::new(instance), bytes)
+    }
+    pub fn call(
+        &self,
+        function: &str,
+        args: &mut typst::eval::Args,
+    ) -> SourceResult<typst::eval::Value> {
+        let span = args.span;
+        let plugin = &mut self.0.lock().unwrap();
+        let ty = plugin
+            .get_function(function)
+            .ok_or("Plugin doesn't have the method: {function}")
+            .at(span)?
+            .ty(plugin.get_store());
+        let arg_count = ty.params().len();
+        let mut str_args = vec![];
+        for k in 0..arg_count {
+            let arg = args
+                .eat::<typst::eval::Value>()?
+                .ok_or(format!("plugin methods takes {arg_count} args, {k} provided"))
+                .at(span)?
+                .cast::<String>()
+                .at(span)?;
+            str_args.push(arg);
+        }
+        let s = plugin
+            .call(
+                function,
+                str_args.iter().map(|x| x.as_str()).collect::<Vec<_>>().as_slice(),
+            )
+            .at(span)?;
+        Ok(typst::eval::Value::Str(s.into()))
+    }
+
+    pub fn has_function(&self, method: &str) -> bool {
+        self.0.lock().unwrap().has_function(method)
+    }
+
+    pub fn iter_func(&self) -> Vec<EcoString> {
+        self.0
+            .lock()
```

Comment:
I don't think the Mutex actually needs to lock the function metadata. What I would probably do is the following: `pub struct Plugin(Arc<Repr>)`, `struct Repr { data: Bytes, functions: Vec<(EcoString, wasmi::Function)>, store: Mutex<wasmi::Store<Storage>> }`.

With these changes, we'd have just one layer of functions instead of two layers and `functions(&self)` could return `impl Iterator<Item = &str>`, which could also make `has_function` unnecessary.